py/desisim/scripts/quickquasars.py

```diff
@@ -382,32 +431,37 @@ def simulate_one_healpix(ifilename,args,model,obsconditions,decam_and_wise_filte
         
     # Redshift distribution resample to match the one in file give by args.dn_dzdm for each type of raw mock
     # Ohio mocks don't have a downsampling.
-    if args.dn_dzdm is not None:
-        if args.downsampling or args.eboss:
-            raise ValueError("dn_dzdm option can not be run with downsampling or eboss options")
+    if args.dn_dzdm:
+        # TODO: Deprecate this option. It will be faster to do this on preprocessing.
+        if args.eboss:
+            raise ValueError("dn_dzdm option can not be run with eboss options")
         dndzdm_file=os.path.join(os.path.dirname(desisim.__file__),'data/dn_dzdM_EDR.fits')
         hdul_dn_dzdm=pyfits.open(dndzdm_file)
         zcenters=hdul_dn_dzdm['Z_CENTERS'].data
         dz = 0.5*(zcenters[1]-zcenters[0]) # Get bin size of the distribution
-
-        if args.dn_dzdm!='ohio':
-            log.info("Resampling to redshift distribution for {} mocks".format(args.dn_dzdm))
-            rnd_state = np.random.get_state()
-            fraction=hdul_dn_dzdm['FRACTIONS_{}'.format(args.dn_dzdm.upper())].data
-            z=metadata['Z']
-
-            zmin = zcenters[0] - dz
-            zmax = zcenters[-1] + dz
-
-            bins = ((z - zmin)/(zmax - zmin) * len(zcenters) + 0.5).astype(np.int64)
-            selection_z = np.random.uniform(size=z.size) < fraction[bins]
-
-            np.random.set_state(rnd_state)
-            log.info("Resampling redshift distribution {}->{}".format(len(z),len(selection_z)))
-
-            transmission = transmission[selection_z]
-            metadata = metadata[:][selection_z]
-            DZ_FOG = DZ_FOG[selection_z]
+        
+        if args.from_catalog is None:
+            if args.raw_mock is None:
+                raise ValueError("If --dn_dzdm option is used the flag --raw-mock must be provided")
+            if args.raw_mock!='ohio':
+                log.info("Resampling to redshift distribution for {} mocks".format(args.raw_mock))
+                rnd_state = np.random.get_state()
+                fraction=hdul_dn_dzdm['FRACTIONS_{}'.format(args.raw_mock.upper())].data
+                z=metadata['Z']
+
+                zmin = zcenters[0] - dz
+                zmax = zcenters[-1] + dz
+
+                bins = ((z - zmin)/(zmax - zmin) * len(zcenters) + 0.5).astype(np.int64)
+                selection_z = np.random.uniform(size=z.size) < fraction[bins]
+
+                np.random.set_state(rnd_state)
```

Comment:
Why do you restore the old random state?

Reply:
I used this to restore the quickquasars state to what it would be if we don't downsample to match the redshift distribution i.e not passing the `--dn_dzdm` flag. This is probably not needed since in either case the realizations would differ in number of objects and therefore the all the random calls done by the code would be affected. So I think It can be removed if requested.